File: 100/107 - Binary Tree Level Order Traversal II.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        Queue<TreeNode> curr = new LinkedList<TreeNode>(), children = new LinkedList<TreeNode>(), temp;
        TreeNode node;
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        List<Integer> vals;
        
        children.add(root);
        while (children.peek() != null) {
            temp = curr;
            curr = children;
            children = temp;
            vals = new ArrayList<Integer>();
            while (! curr.isEmpty()) {
                node = curr.poll();
                vals.add(node.val);
                if (node.left != null)
                    children.add(node.left);
                if (node.right != null)
                    children.add(node.right);
            }
            ret.add(0, vals);
        }
        
        return ret;
    }
}

File: csharp/107.binary-tree-level-order-traversal-ii.cs
/*
 * @lc app=leetcode id=107 lang=csharp
 *
 * [107] Binary Tree Level Order Traversal II
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */

using System.Collections.Generic;

public class Solution {
    public IList<IList<int>> LevelOrderBottom(TreeNode root) {
        var ret = new List<IList<int>>();
        var queue = new List<TreeNode>();
        int count;
        queue.Add(root);
        while ((count = queue.Count) != 0) {
            var add = new List<int>();
            while (count-- > 0) {
                var node = queue[0];
                queue.RemoveAt(0);
                if (node == null) {
                    continue;
                }
                queue.Add(node.left);
                queue.Add(node.right);
                add.Add(node.val);
            }
            if (add.Count > 0) {
                ret.Add(add);
            }
        }
        ret.Reverse();
        return ret;
    }
}

